hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java

```diff
@@ -878,6 +879,24 @@ private synchronized void addResourceObject(Resource resource) {
   private static final int SUB_START_IDX = 0;
   private static final int SUB_END_IDX = SUB_START_IDX + 1;
 
+  private Set<String> PROXY_USERS = null;
+
+  public synchronized Set<String> getProxyUsers() {
+    if (PROXY_USERS == null) {
+      PROXY_USERS = new HashSet<>();
+      for (Map.Entry<String, String> entry : this) {
+        String propName = entry.getKey();
+        if (propName.startsWith(ProxyUsers.CONF_HADOOP_PROXYUSER)) {
+          String[] tokens = propName.split("\\.");
+          // Configuration property is in the form of hadoop.proxyuser.USERNAME.{hosts,groups}
+          PROXY_USERS.add(tokens[2]);
```

Comment:
Since we are refactoring it, we could be a little bit more defensive. Either check the property name with a regex and extract the group we are interested or check if the `tokens` array has indeed 2+ elements.